posts/inside-rust/2024-08-09-async-closures-call-for-testing.md

```diff
@@ -0,0 +1,124 @@
+
+Somewhat relatedly, on the callee side, when users want to take an async closure as an argument, they must express that as a bound of two different generic types, or use boxing:
+
+```rust
+fn async_callback<F, Fut>(callback: F)
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = String>;
+
+// Or:
+fn async_callback_boxed<F, Fut>(callback: F)
+where
+    F: FnOnce() -> Pin<Box<dyn Future<Output = String>>>;
+```
+
+This led to an additional limitation, that it's impossible to express higher-ranked async fn bounds without boxing, since a higher-ranked trait bound on `F` cannot lead to a higher-ranked type for `Fut`.
+
+These limitations were detailed in [Niko's blog post on async closures and lending](https://smallcultfollowing.com/babysteps/blog/2023/05/09/giving-lending-and-async-closures/#async-closures-are-a-lending-pattern), and later in compiler-errors's blog post on [why async closures are the way they are](https://hackmd.io/@compiler-errors/async-closures).
+
+## OK, so how does [RFC 3668] help?
+
+Recent [work](https://github.com/rust-lang/rust/pull/120361) has focused on reimplementing async closures to be lending and designing a set of async fn traits. While async closures already existed as syntax, this work introduced a new family of async fn traits which are implemented by async closures (and all other callable types which return futures). They can be written like:
+
+```rust
+fn test<F>(callback: F)
+where
+    // Either:
+    async Fn(Arg, Arg) -> Ret
+    // Or:
+    AsyncFn(Arg, Arg) -> Ret
+```
+
+(It's currently an [open question](https://github.com/rust-lang/rust/issues/128129) exactly how to spell this bound, so both syntaxes are implemented in parallel.)
+
+RFC 3668 motivates this implementation work in detail, confirming that we need first-class async closures and async fn traits which allow us to express the *lending* capability of async closures -- read the RFC if you're interested in the whole story!
+
+## So how do I help?
+
+We'd love for you to test out these new features! First, on a recently-updated nightly compiler, enable `#![feature(async_closure)]` (note that, for historical reasons, this feature name is not pluralized).
+
+Async closures are designed to be drop-in compatible (in almost all cases) with closures returning async blocks:
+
+```rust
+// Instead of writing:
+takes_async_callback(|arg| async {
+    // Do things here...
+});
+
+// Write this:
+takes_async_callback(async |arg| {
+    // Do things here...
+});
+```
+
+And on the callee side, write async fn trait bounds instead of writing "regular" fn trait bounds that return futures:
+
+```rust
+// Instead of writing:
+fn doesnt_exactly_take_an_async_closure<F, Fut>()
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = String>
+{ todo!() }
+
+// Write this:
+fn takes_an_async_closure<F: async FnOnce() -> String>() { todo!() }
+// Or this:
+fn takes_an_async_closure<F: AsyncFnOnce() -> String>() { todo!() }
+```
```

Comment:
Show an example of rewriting `Fn(&T) -> Box<dyn Future + '_>` to a higher ranked async fn bound